### PR TITLE
Introduce visitor-based parallel rule runner (Rayon)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
Adds a new visitor-based parallel rule runner that dispatches CST rule evaluations across a Rayon thread pool. This completes the core linter engine by connecting rule store, context, and runner with per-rule parallelism while preserving deterministic diagnostic output.

## Changes
- Add dependency: rayon = "1.10" in Cargo.toml.
- Add src/linter/runner.rs implementing:
  - Runner<'a> struct storing store, GreenNode, source text, and RuleConfig.
  - Runner::new(store, source_text, parsed, config) -> Self.
  - Runner::run() -> Vec<LintDiagnostic> with per-rule parallelism and deterministic sorting.
  - Internal helper run_single_rule and sort_diagnostics.
  - Unit tests validating Send + Sync, empty store behavior, single/multiple rules, empty target kinds, config propagation, and deterministic ordering.
- Wire Runner into linter module:
  - Update src/linter/mod.rs to declare mod runner; and pub use runner::Runner.
- Add behavioural tests in tests/linter_runner.rs to exercise Runner against existing store/rules and compare to sequential dispatch baseline.
- Documentation updates:
  - docs/ddlint-design.md: document Runner API contract, per-rule parallelism, and thread-safety rationale.
  - docs/execplans/3-1-4-visitor-based-parallel-rule-runner.md: new ExecPlan describing the runner implementation (Stage A–F, testing plan, risks, etc.).
  - docs/roadmap.md: mark 3.1.4 as done.
  - docs/execplans/3-1-3-cst-rule-store.md: minor wording alignment about runner context.
- Executed quality gates: formatting, linting, tests, and markdown lintments supported by the extended test suite and docs.

## Implementation details
- Per-rule parallelism: Since rowan SyntaxNode/SyntaxToken are !Send + !Sync, the runner clones a shared GreenNode per rayon task and rebuilds a thread-local red tree via Root::from_green(green.clone()), enabling safe parallel rule evaluation.
- API surface:
  - Runner<'a> {
      store: &'a CstRuleStore,
      green: GreenNode,
      source_text: Arc<str>,
      config: RuleConfig,
    }
  - pub fn new(store, source_text, parsed, config) -> Self
  - pub fn run(&self) -> Vec<LintDiagnostic>
- Diagnostics are collected per-rule, flattened, and deterministically sorted by (span.start, span.end, rule_name).
- Runner::new() takes &Parsed to extract the GreenNode, matching existing patterns and ergonomics.

## Tests
- Unit tests (src/linter/runner.rs) include:
  - runner_is_send_and_sync: ensures Runner is Send + Sync.
  - empty_store_produces_no_diagnostics
  - single_rule_produces_diagnostics
  - diagnostics_are_sorted_by_span
  - multiple_rules_produce_merged_diagnostics
- Behavioural tests (tests/linter_runner.rs) cover:
  - runner_matches_sequential_dispatch: compares parallel Runner results against a sequential baseline across hello_join.dl and reachability.dl fixtures.
  - runner_with_empty_store_produces_no_diagnostics
  - runner_with_multiple_rules
  - runner_handles_empty_target_kinds
  - runner_respects_config
  - runner_produces_deterministic_ordering
  - runner_diagnostics_sorted_by_span_then_rule_name

## Documentation & Roadmap
- docs/ddlint-design.md: Runner API contract and per-rule parallelism rationale added.
- docs/execplans/3-1-4-visitor-based-parallel-rule-runner.md: new ExecPlan detailing stages, tests, and risks.
- docs/roadmap.md: 3.1.4 marked as done.
- docs/execplans/3-1-3-cst-rule-store.md: minor wording tweak to reflect runner integration.

## Validation plan
- Run: make check-fmt, make lint, make test, and make markdownlint to ensure all gates pass.
- Validate that diagnostics are deterministic and equivalent to the sequential baseline.
- Confirm no public API breakage beyond adding Runner (new type exposed under ddlint::linter).

## Notes
- Existing parsing and rule definitions remain unchanged; this work focuses on parallel execution orchestration and tests.
- The per-rule design constraints (thread-safety and Send/Sync boundaries) are documented inline and in design docs to guide future rule development.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/6072442f-ac08-4339-807c-46293b5acdf5

## Summary by Sourcery

Introduce a visitor-based parallel linter runner that executes CST rules over a Rayon thread pool and exposes it via the linter module.

New Features:
- Add a public Runner type that traverses the CST and executes registered CST rules in parallel with deterministic diagnostic ordering.

Enhancements:
- Document the Runner API and per-rule parallelism strategy in the design docs and add an execution plan describing the implementation stages, risks, and validation.
- Update the linter module docs to describe the full engine including the parallel runner and adjust wording in the CST rule store exec plan.

Build:
- Add the rayon crate as a dependency for parallel rule execution.

Documentation:
- Extend ddlint design documentation with the Runner contract and thread-safety rationale and mark roadmap item 3.1.4 as complete.
- Add an execplan document for the visitor-based parallel rule runner, including context, constraints, risks, progress, and validation steps.

Tests:
- Add unit tests for the Runner covering Send/Sync, empty and populated stores, diagnostic sorting, and multi-rule scenarios, plus behavioural tests comparing parallel execution to a sequential baseline and validating determinism and config handling.

### CODE REVIEW

File: tests/linter_runner/main.rs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+             before ({:?}, {:?}, {:?})",
+            a.span().start(),
+            a.span().end(),
+            a.rule_name(),
+            b.span().start(),
+            b.span().end(),
+            b.rule_name(),
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/linter_runner/main.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
